src/libraries/System.Private.CoreLib/src/System/SpanHelpers.cs

```diff
@@ -332,77 +333,129 @@ public static unsafe void ClearWithoutReferences(ref byte b, nuint byteLength)
+                {
+                    nuint stopLoopAtOffset = totalByteLength & (nuint)(nint)(2 * -Vector<byte>.Count); // intentional sign extension carries the negative bit
+
+                    do
+                    {
+                        Unsafe.WriteUnaligned(ref Unsafe.AddByteOffset(ref refDataAsBytes, offset), zero);
+                        Unsafe.WriteUnaligned(ref Unsafe.AddByteOffset(ref refDataAsBytes, offset + (nuint)Vector<byte>.Count), zero);
+                        offset += (uint)(2 * Vector<byte>.Count);
+                    } while (offset < stopLoopAtOffset);
+                }
+
+                // At this point, if any data remains to be written, it's strictly less than
+                // 2 * sizeof(Vector) bytes. The loop above had us write an even number of vectors.
+                // If the total byte length instead involves us writing an odd number of vectors, write
+                // one additional vector now. The bit check below tells us if we're in an "odd vector
+                // count" situation.
+
+                if ((totalByteLength & (nuint)Vector<byte>.Count) != 0)
+                {
+                    Unsafe.WriteUnaligned(ref Unsafe.AddByteOffset(ref refDataAsBytes, offset), zero);
+                }
+
+                // It's possible that some small buffer remains to be populated - something that won't
+                // fit an entire vector's worth of data. Instead of falling back to a loop, we'll write
+                // a vector at the very end of the buffer. This may involve overwriting previously
+                // populated data, which is fine since we're just zeroing everything out anyway.
+                // There's no need to perform a length check here because we already performed this
+                // check before entering the vectorized code path.
+
+                Unsafe.WriteUnaligned(ref Unsafe.AddByteOffset(ref refDataAsBytes, totalByteLength - (nuint)Vector<byte>.Count), zero);
             }
             else
             {
-                return; // nothing to write
-            }
+                // If we reached this point, vectorization is disabled, or there are too few
+                // elements for us to vectorize. Fall back to an unrolled loop.
+
+                nuint i = 0;
 
-        Write4To7:
-            Debug.Assert(pointerSizeLength >= 4);
+                // Write 8 elements at a time
+                // Skip this check if "write 8 elements" would've gone down the vectorized code path earlier
 
-            // Write first four and last three.
-            Unsafe.Add(ref ip, 2) = default;
-            Unsafe.Add(ref ip, 3) = default;
-            Unsafe.Add(ref Unsafe.Add(ref ip, (nint)pointerSizeLength), -3) = default;
-            Unsafe.Add(ref Unsafe.Add(ref ip, (nint)pointerSizeLength), -2) = default;
+                if (!Vector.IsHardwareAccelerated || Vector<byte>.Count / sizeof(IntPtr) > 8) // JIT turns this into constant true or false
```

Comment:
I'm missing it... why is this check necessary?

Reply:
JIT won't properly elide the following `if` without this check. Here's the codegen if this and the similar check a few lines down are removed. The fallback logic becomes _much_ larger. But note that blocks `d07b` - `d0ba` and blocks `d0c1` - `d0e2` will never be executed. The explicit `if` in the code above is hinting to JIT that it shouldn't worry about this scenario. (That's also the crux behind the "skip this check if..." comments above these lines, but I can wordsmith if necessary.)

```asm
00007ffa`30eed000 c5f877                 vzeroupper 
00007ffa`30eed003 4883fa04               cmp     rdx, 4
00007ffa`30eed007 726a                   jb      System_Private_CoreLib!System.SpanHelpers.ClearWithReferences(IntPtr ByRef,  UIntPtr)+0x73 (00007ffa`30eed073)

;; BEGIN MAIN VECTOR LOGIC
00007ffa`30eed009 c4e17c57c0             vxorps  ymm0, ymm0, ymm0
00007ffa`30eed00e c4e17d1101             vmovupd ymmword ptr [rcx], ymm0
; <snip>
00007ffa`30eed06f c5f877                 vzeroupper 
00007ffa`30eed072 c3                     ret     
;; END MAIN VECTOR LOGIC

;; BEGIN FALLBACK LOGIC
00007ffa`30eed073 33c0                   xor     eax, eax
00007ffa`30eed075 4883fa08               cmp     rdx, 8  ; rdx can never be < 8 (see lines d003 - d007 earlier)
00007ffa`30eed079 7241                   jb      System_Private_CoreLib!System.SpanHelpers.ClearWithReferences(IntPtr ByRef,  UIntPtr)+0xbc (00007ffa`30eed0bc)
00007ffa`30eed07b 4c8bc2                 mov     r8, rdx
00007ffa`30eed07e 4983e0f8               and     r8, 0FFFFFFFFFFFFFFF8h
00007ffa`30eed082 4c8bc8                 mov     r9, rax
00007ffa`30eed085 49c1e103               shl     r9, 3
00007ffa`30eed089 4533d2                 xor     r10d, r10d
00007ffa`30eed08c 4e891409               mov     qword ptr [rcx+r9], r10
00007ffa`30eed090 4e89540908             mov     qword ptr [rcx+r9+8], r10
00007ffa`30eed095 4e89540910             mov     qword ptr [rcx+r9+10h], r10
00007ffa`30eed09a 4e89540918             mov     qword ptr [rcx+r9+18h], r10
00007ffa`30eed09f 4e89540920             mov     qword ptr [rcx+r9+20h], r10
00007ffa`30eed0a4 4e89540928             mov     qword ptr [rcx+r9+28h], r10
00007ffa`30eed0a9 4e89540930             mov     qword ptr [rcx+r9+30h], r10
00007ffa`30eed0ae 4e89540938             mov     qword ptr [rcx+r9+38h], r10
00007ffa`30eed0b3 4883c008               add     rax, 8
00007ffa`30eed0b7 493bc0                 cmp     rax, r8
00007ffa`30eed0ba 72c6                   jb      System_Private_CoreLib!System.SpanHelpers.ClearWithReferences(IntPtr ByRef,  UIntPtr)+0x82 (00007ffa`30eed082)
00007ffa`30eed0bc f6c204                 test    dl, 4  ; rdx & 4 will always result in 0 (see lines d003 - d007 earlier)
00007ffa`30eed0bf 7421                   je      System_Private_CoreLib!System.SpanHelpers.ClearWithReferences(IntPtr ByRef,  UIntPtr)+0xe2 (00007ffa`30eed0e2)
00007ffa`30eed0c1 4c8bc0                 mov     r8, rax
00007ffa`30eed0c4 49c1e003               shl     r8, 3
00007ffa`30eed0c8 4533c9                 xor     r9d, r9d
00007ffa`30eed0cb 4e890c01               mov     qword ptr [rcx+r8], r9
00007ffa`30eed0cf 4e894c0108             mov     qword ptr [rcx+r8+8], r9
00007ffa`30eed0d4 4e894c0110             mov     qword ptr [rcx+r8+10h], r9
00007ffa`30eed0d9 4e894c0118             mov     qword ptr [rcx+r8+18h], r9
00007ffa`30eed0de 4883c004               add     rax, 4
00007ffa`30eed0e2 f6c202                 test    dl, 2
00007ffa`30eed0e5 7417                   je      System_Private_CoreLib!System.SpanHelpers.ClearWithReferences(IntPtr ByRef,  UIntPtr)+0xfe (00007ffa`30eed0fe)
00007ffa`30eed0e7 4c8bc0                 mov     r8, rax
00007ffa`30eed0ea 49c1e003               shl     r8, 3
00007ffa`30eed0ee 4533c9                 xor     r9d, r9d
00007ffa`30eed0f1 4e890c01               mov     qword ptr [rcx+r8], r9
00007ffa`30eed0f5 4e894c0108             mov     qword ptr [rcx+r8+8], r9
00007ffa`30eed0fa 4883c002               add     rax, 2
00007ffa`30eed0fe f6c201                 test    dl, 1
00007ffa`30eed101 7406                   je      System_Private_CoreLib!System.SpanHelpers.ClearWithReferences(IntPtr ByRef,  UIntPtr)+0x109 (00007ffa`30eed109)
00007ffa`30eed103 33d2                   xor     edx, edx
00007ffa`30eed105 488914c1               mov     qword ptr [rcx+rax*8], rdx
00007ffa`30eed109 c5f877                 vzeroupper 
00007ffa`30eed10c c3                     ret
```